### DISCUSSION TITLE
How to unit test a parallel-consumer using MockConsumer?

### DISCUSSION THREAD

**Discussion by cykl** [confluentinc/parallel-consumer#512] [Q&A]
parallel-consumer-example-core is nice because it provides a minimal example of a parallel-consumer based application and some unit tests. However, tests rely on LongPollingMockConsumer which is not available to parallel-consumer users. 

It degrades developer experience as one has to copy LongPollingMockConsumer (which isn't pleasant because it depends on other classes and Lombock) or to study it to understand what is required an what is not .

After spending some time trying to write an unit test it is still unclear to me why PC is complaining that epochOfPartition is null and will have to troubleshoot it. 

```java
    @Test
    void test0( ) throws InterruptedException {
        var topicName = "topic1";

        var record = new ConsumerRecord<>( topicName, 0, 0, "key1", "value1" );

        var mockConsumer = new MockConsumer<String, String>( OffsetResetStrategy.EARLIEST );
        mockConsumer.schedulePollTask( ( ) -> {
            var topicPartition = new TopicPartition( topicName, 0 );
            mockConsumer.rebalance( List.of( topicPartition ) );
            mockConsumer.updateBeginningOffsets( Map.of( topicPartition, 0L ) );
            mockConsumer.addRecord( record );
        } );

        try (mockConsumer) {
            var myConsumer = new MyConsumer( mockConsumer );
            myConsumer.startBySubscribing( Pattern.compile( "topic1" ) );
           
            // Do something here
      }
}
```

It would be much nicer to either have a ready to use mock consumer or some examples that can be reused.  

As a short term answer to my specific question, could you share a minimal unit test based on  MockConsumer?

**Comment by astubbs** [comment]
> However, tests rely on LongPollingMockConsumer which is not available to parallel-consumer users.

It is, however it lives in the test artifice from core, so you have to add that as a dep to access it.  However, in 0.6, we're moving it into the main code, so it's easier. FYI created to track:
- #526

To import the test dep, use:
https://github.com/confluentinc/parallel-consumer/blob/a5b3409a211693b731ce35334067f289b41e995e/parallel-consumer-examples/parallel-consumer-example-core/pom.xml#L27-L33

Also, check out this comment out. You can use MockConsumer as well, but it's a PITA (hence LongPollingMockConsumer)
https://github.com/confluentinc/parallel-consumer/issues/176#issuecomment-1302129191

MockConsumer example - although I suggest you import the test dep from core, and use LongPollingMockConsumer:
https://github.com/confluentinc/parallel-consumer/pull/470

Are you able to find some examples of using LongPollingMockConsumer with PC? If not, I can point them out, or write a couple minimalistic ones.

And finally - thank you VERY much for your help with the metrics stuff - much appreciated! Really looking forward to getting that in.

Please reach out if you have more questions! :)

**Comment by astubbs** [comment]
@cykl - the fix you asked for is released: `LongPollingMockConsumer` ships in the main `parallel-consumer-core` artefact on [v0.6.0.0](https://github.com/astubbs/parallel-consumer/releases/tag/v0.6.0.0) of the maintained fork, so a broker-free test needs no `tests` classifier jar and no copy (astubbs/parallel-consumer#202; this tracker's #526, mirrored as astubbs/parallel-consumer#159). It is `bz.stub.parallelconsumer.internal.utils.LongPollingMockConsumer`.

On the `epochOfPartition is null` you hit with a bare `MockConsumer`: the epoch is set by the assignment callback, so a record added before `rebalance(...)` has run through PC's listener reaches the engine for a partition it does not know yet. `LongPollingMockConsumer.subscribeWithRebalanceAndAssignment(topics, partitions)` does the subscribe, the rebalance and the assignment in the right order, which is why it exists; `CoreAppTest` in `parallel-consumer-example-core` is the minimal example, and the Vert.x, Reactor and metrics example tests use the same pattern. On v0.6.0.0 that null is also no longer fatal - a record polled for a partition whose callback has not run is skipped and redelivered rather than raising.

An open pull request on the fork, not yet merged: a broker-free sandbox module that pipes records into either API and reads what came out (astubbs/parallel-consumer#507, open), which is the ready-to-use harness you described.

The coordinates changed with the fork - on Maven Central it is now:

```xml
<dependency>
    <groupId>bz.stub.parallelconsumer</groupId>
    <artifactId>parallel-consumer-core</artifactId>
    <version>0.6.0.0</version>
</dependency>
```

(`parallel-consumer-vertx`, `parallel-consumer-reactor` and `parallel-consumer-mutiny` likewise.) The Java package is `bz.stub.parallelconsumer` too, so imports change with it - one command rewrites them (GNU `sed`; on macOS use `sed -i ''`):

```bash
find . -name '*.java' -exec sed -i 's/io\.confluent\.parallelconsumer/bz.stub.parallelconsumer/g' {} +
```
